conan/internal/runner/docker.py

```diff
@@ -0,0 +1,159 @@
+                    detach=True,
+                    auto_remove=False)
+                self.init_container()
+            self.run_command(self.command)
+            self.update_local_cache()
+        except:
+            ConanOutput().error(f'Something went wrong running "{self.command}" inside the container')
+        finally:
+            if self.container:
+                docker_info('Stopping container')
+                self.container.stop()
+                if self.remove:
+                    docker_info('Removing container')
+                    self.container.remove()
+
+    def build_image(self):
+        if self.dockerfile:
+            docker_build_logs = self.docker_api.build(path=self.dockerfile, tag=self.image)
+            for chunk in docker_build_logs:
+                for line in chunk.decode("utf-8").split('\r\n'):
+                    if line:
+                        stream = json.loads(line).get('stream')
+                        if stream:
+                            ConanOutput().info(stream.strip())
+
+    def run_command(self, command, log=True, stream=True, tty=True):
+        if log:
+            docker_info(f'Running in container: "{command}"')
+        exec_run = self.container.exec_run(f"/bin/bash -c '{command}'", stream=stream, tty=tty)
+        if stream:
+            try:
+                while True:
+                    chunk = next(exec_run.output).decode('utf-8', errors='ignore').strip()
+                    if log:
+                        ConanOutput().info(chunk)
+            except StopIteration:
+                pass
+        else:
+            return exec_run.output.decode('utf-8', errors='ignore').strip()
+
+    def create_runner_environment(self):
+        volumes = {self.abs_host_path: {'bind': self.abs_docker_path, 'mode': 'rw'}}
+        environment = {'CONAN_RUNNER_ENVIRONMENT': '1'}
+        if self.cache == 'shared':
+            volumes[ConfigAPI(self.conan_api).home()] = {'bind': '/root/.conan2', 'mode': 'rw'}
+        if self.cache in ['clean', 'copy']:
+            shutil.rmtree(self.abs_runner_home_path, ignore_errors=True)
+            os.mkdir(self.abs_runner_home_path)
+            os.mkdir(os.path.join(self.abs_runner_home_path, 'profiles'))
+            for file_name in ['global.conf', 'settings.yml', 'remotes.json']:
+                src_file = os.path.join(ConfigAPI(self.conan_api).home(), file_name)
+                if os.path.exists(src_file):
+                    shutil.copy(src_file, os.path.join(self.abs_runner_home_path, file_name))
+            self._copy_profiles(self.args.profile_build)
+            self._copy_profiles(self.args.profile_host)
+            if self.cache == 'copy':
+                tgz_path = os.path.join(self.abs_runner_home_path, 'local_cache_save.tgz')
+                docker_info(f'Save host cache in: {tgz_path}')
+                self.conan_api.cache.save(self.conan_api.list.select(ListPattern("*:*")), tgz_path)
```

Comment:
Im thinking now, why did we go for copying from the host cache to docker? In which cases does this help? Im thinking most of the time (Except for header-only recipes for example) the packages will be different and thus will still be needed from a remote.

Is it to get exported recipes there that the user might be developing?

Reply:
if you set `cache=clean` it doesn't copy anything. It can be useful in case you use docker inside linux or you use the output of a docker create in another one.